### PR TITLE
feat: [CDS-51185] Terraform Cloud manager side run step logic

### PR DESCRIPTION
**Pull request by alebuh**
## Describe your changes
Manager side logic for executing Terraform Cloud Run step
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43230)

### CODE REVIEW

**Comment by alebuh** [comment]
trigger smartchecks

File: 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudParamsMapper.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    if (isEmpty(inputVariables)) {
+      return new HashMap<>();
+    }
+    Map<String, String> res = new LinkedHashMap<>();
+    inputVariables.keySet().forEach(
+        key -> res.put(key, ((ParameterField<?>) inputVariables.get(key)).getValue().toString()));
+    return res;
+  }
+
+  private io.harness.delegate.beans.terraformcloud.PlanType getPlanType(PlanType planType) {
```
Comment (vlica-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudParamsMapper.java:131]: is sonar not raising a concern regarding such refference here?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudStepHelper.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    String fullEntityId = generateFullIdentifier(
+        ParameterFieldHelper.getParameterFieldValue(planSpecParameters.getProvisionerIdentifier()), ambiance);
+    String outputName = format(TFC_OUTPUT_FORMAT, PLAN.getDisplayName(), fullEntityId);
+    executionSweepingOutputService.consume(
+        ambiance, outputName, terraformCloudPlanOutput, StepOutcomeGroup.STAGE.name());
+  }
+
+  public String getTerraformPlanName(PlanType planType, Ambiance ambiance, String provisionId) {
+    String prefix = PlanType.APPLY == planType ? TFC_PLAN_NAME_PREFIX_NG : TFC_DESTROY_PLAN_NAME_PREFIX_NG;
+    return format(prefix, ambiance.getPlanExecutionId(), provisionId).replace("_", "-");
```
Comment (vlica-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudStepHelper.java:126]: is this replace from _ to - required?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudStepHelper.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+  }
+
+  public String getProvisionIdentifier(TerraformCloudRunSpecParameters runSpec) {
+    ParameterField<String> provisionerIdentifier;
+    switch (runSpec.getType()) {
+      case REFRESH_STATE:
+        return null;
```
Comment (vlica-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudStepHelper.java:133]: Do we check later if provisionerIdenfier is null? I believe we save this property in DB?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/outcome/TerraformCloudRunOutcome.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import io.harness.pms.sdk.core.data.Outcome;
+
+import java.util.HashMap;
+import lombok.Builder;
+import lombok.Data;
+
+@OwnedBy(HarnessTeam.CDP)
+@Data
+@Builder
+@RecasterAlias("io.harness.cdng.provision.terraformcloud.outcome.TerraformCloudRunOutcome")
```
Comment (vlica-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/outcome/TerraformCloudRunOutcome.java:22]: Is this alias added in alias.txt

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/terraformcloud/TerraformCloudTaskType.java [hunk 5]
```diff
@@ -5,9 +5,20 @@
...
+@OwnedBy(HarnessTeam.CDP)
+public enum TerraformCloudTaskType {
+  VALIDATE,
+  GET_ORGANIZATIONS,
+  GET_WORKSPACES,
+  RUN_REFRESH_STATE,
+  RUN_PLAN_ONLY,
+  RUN_PLAN_AND_APPLY,
+  RUN_PLAN_AND_DESTROY,
+  RUN_PLAN,
```
Comment (vlica-harness) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/terraformcloud/TerraformCloudTaskType.java:22]: just curios, what's the differnce between RUN_PLAN_ONLY vs RUN_PLAN?

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudCommandUnit.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+ * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
+ */
+
+package io.harness.delegate.task.terraformcloud;
+
+import static io.harness.annotations.dev.HarnessTeam.CDP;
+
+import io.harness.annotations.dev.OwnedBy;
+
+@OwnedBy(CDP) public enum TerraformCloudCommandUnit { Run }
```
Comment (vlica-harness) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudCommandUnit.java:14]: should Run be capital case?

**Review by alebuh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43230)* status: 26 file(s) to review. 6. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudParamsMapper.java` line 131 at r2](https://reviewable.io/reviews/harness/harness-core/43230#-NN21qIlPSnP7EfvkB64-r2-131:-NNCeJFIEKkXgFJD6mCp:b6vrv4r) ([raw file](https://github.com/harness/harness-core/blob/de074c0ed51b407446f506b0ad0bddde3a93e137/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudParamsMapper.java#L131)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

is sonar not raising a concern regarding such refference here?
</blockquote></details>

Sonar is fine with this. We have two classes with the same name used here, that's why we have this reference 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudStepHelper.java` line 126 at r2](https://reviewable.io/reviews/harness/harness-core/43230#-NN21qIlPSnP7EfvkB65-r2-126:-NNCm2xP87zpaF2E7TyA:bs94np3) ([raw file](https://github.com/harness/harness-core/blob/de074c0ed51b407446f506b0ad0bddde3a93e137/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudStepHelper.java#L126)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

is this replace from _ to - required? 
</blockquote></details>

I saw it was done for Terraform and Terragrunt so I did the same. Not sure if is it required or not but I assume that there is a reason why it was done.
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudStepHelper.java` line 133 at r2](https://reviewable.io/reviews/harness/harness-core/43230#-NN21qIlPSnP7EfvkB65-r2-133:-NNCmUdT5_mrMoXBl0To:b-ygvz3k) ([raw file](https://github.com/harness/harness-core/blob/de074c0ed51b407446f506b0ad0bddde3a93e137/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudStepHelper.java#L133)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

Do we check later if provisionerIdenfier is null? I believe we save this property in DB?
</blockquote></details>

Actually, we should never call this method when the type is REFRESH_STATE, therefore I'm removing that branch assuming that default will throw an exception if it happens. (There isn't provisionerId for REFRESH_STATE) 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/outcome/TerraformCloudRunOutcome.java` line 22 at r2](https://reviewable.io/reviews/harness/harness-core/43230#-NN21qIlPSnP7EfvkB66-r2-22:-NNCms2Y3a4IG2ibkr3X:bj63v1s) ([raw file](https://github.com/harness/harness-core/blob/de074c0ed51b407446f506b0ad0bddde3a93e137/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/outcome/TerraformCloudRunOutcome.java#L22)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

Is this alias added in alias.txt
</blockquote></details>

added now!
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/terraformcloud/TerraformCloudTaskType.java` line 22 at r2](https://reviewable.io/reviews/harness/harness-core/43230#-NN6Tv2sgN1RuVpUwjHl-r2-22:-NNCmvXgE6K-EA0eL6AA:b-ebp4xi) ([raw file](https://github.com/harness/harness-core/blob/de074c0ed51b407446f506b0ad0bddde3a93e137/950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/terraformcloud/TerraformCloudTaskType.java#L22)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

just curios, what's the differnce between RUN_PLAN_ONLY vs RUN_PLAN?
</blockquote></details>

RUN_PLAN_ONLY is Plan that can't be applied, while RUN_PLAN should be applied or declined after it was created.
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudCommandUnit.java` line 14 at r2](https://reviewable.io/reviews/harness/harness-core/43230#-NN21qIlPSnP7EfvkB6A-r2-14:-NNCnFdv1KGEMD_b_Wx8:b-n9tkhd) ([raw file](https://github.com/harness/harness-core/blob/de074c0ed51b407446f506b0ad0bddde3a93e137/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudCommandUnit.java#L14)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

should Run be capital case?
</blockquote></details>

changed

**Review by vlica-harness — APPROVED** [review]
> Reviewed 17 of 25 files at r1, 2 of 3 files at r3, 5 of 6 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43230)* status: 26 file(s) to review. 6. unresolved discussions. 1 of 2 approvals obtained.

**Review by vlica-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43230)* status: 26 file(s) to review. 6. unresolved discussions. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 17 of 25 files at r1, 2 of 3 files at r3, 4 of 6 files at r4, 2 of 2 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43230)* status: 6. unresolved discussions.

**Review by vlica-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43230)* status: :shipit: complete! ✓

**Review by tmacari-harness — APPROVED** [review]
> Reviewed 17 of 25 files at r1, 2 of 3 files at r3, 4 of 6 files at r4, 2 of 2 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43230)* status: :shipit: complete! ✓

**Pull request merged**